src/socket/tcp.rs

```diff
@@ -2246,6 +2286,14 @@ impl<'a> Socket<'a> {
             max_seg_size: None,
             sack_permitted: false,
             sack_ranges: [None, None, None],
+            timestamp: if let Some(ts_generator) = self.tsval_generator {
```

Comment:
I think this logic, as well as the logic I highlighted above, should live in an inherent method on `TcpTimestampRepr` which accepts the two timestamps and a TSVal generator and returns one you can use in response or None.